File: sdks/deno/examples/sidecar.ts
// patala SIDECAR mode on Deno — the `patala-sidecar` binary in a child
// process, HTTP over loopback.
//
//   cargo build -p patala-sidecar              # from the repo root
//   PATALA_SIDECAR_BINARY=../../target/debug/patala-sidecar deno task example:sidecar
//   # or, spelled out:
//   deno run --allow-run --allow-net=127.0.0.1 --allow-env examples/sidecar.ts
//
// No --allow-ffi: there is no native code in this path and nothing is loaded
// into this process. Note that --allow-net IS here and IS meaningful — and it
// is scoped to 127.0.0.1, because that is the only host this example may
// legitimately reach. Direct mode's `--allow-ffi` could not make that promise
// even if it wanted to.
//
// The reason to choose the sidecar is KEY ISOLATION: a signing key lives in the
// sidecar's process and nowhere else, instead of in every process of a polyglot
// stack that linked the library.
//
// Everything below runs against MockRail — which is not a choice this example
// gets to make, because the sidecar's registry is mock-only today. That is
// stated rather than hidden: see the 404 near the end.
//
// If you are behind a proxy: Deno's fetch honours HTTP(S)_PROXY for loopback
// URLs too, so export NO_PROXY=127.0.0.1 or this process cannot reach its own
// sidecar and start() times out against a server that is running fine.

import { Sidecar, type SidecarHttpError } from "../mod.ts";

console.log(`deno        ${Deno.version.deno} on ${Deno.build.os}/${Deno.build.arch}`);

// start() mints a 32-byte token and passes it in the child's ENVIRONMENT, never
// in argv — argv is world-readable via `ps`, and this token authorises
// `charge`. The sidecar refuses to start at all without it: there is no
// unauthenticated mode and no generated fallback.
//
// `await using` kills the child and waits for it on the way out, throw
// included; Deno fails a run that leaks one.
await using side = await Sidecar.start({ stdio: "null" });
console.log(`sidecar     ${side.baseURL}  (loopback only, hardcoded — not a knob)`);
console.log(`token       ${side.token.slice(0, 8)}… (32 random bytes, in the environment, not argv)`);
console.log(`healthz     ${await side.healthz()} — the one unauthenticated route\n`);

const caps = await side.capabilities();
const settlement = JSON.stringify(caps.settlement);
console.log(
  `caps        ${caps.class}, currencies ${caps.currencies.join(", ")}, settlement ${settlement}`,
);

const req = {
  amount_minor: 1250,
  currency: "USDC",
  destination: "mock:wallet:alice",
  reference: "order-1",
};

const q = await side.quote(req);
console.log(`quote       ${q.amount_minor} + ${q.fee_minor} fee = ${q.total_minor} ${q.currency}`);

const receipt = await side.charge(req);
console.log(
  `charge      ${receipt.amount_minor} ${receipt.currency} for ${receipt.reference}, proof ${receipt.proof.length} bytes`,
);

// The same JSON direct mode returns — identical wire contract, different
// transport. A body that works here works against patala_call unchanged.
console.log(`verify      ${JSON.stringify(await side.verify(receipt))}`);

// 200 with {"valid": false}, never an HTTP error, so "verified false" can never
// be mistaken for "the sidecar broke".
const tampered = { ...receipt, reference: "order-99" };
console.log(
  `tampered    HTTP 200 ${JSON.stringify(await side.verify(tampered))} — an answer, not a failure\n`,
);

// Every verdict, including the refusals, is a 200. Branch on `status` and
// `is_refusal`, never on the status code.
for (const dest of ["mock:wallet:alice", "mock:program:vault", "nonsense"]) {
  const v = await side.validateDestination(dest);
  const status = v.status.padEnd(18);
  console.log(
    `destination ${dest.padEnd(19)} ${status} refusal=${v.is_refusal} confirm=${v.human_must_confirm}`,
  );
}
console.log();

// The mock rail has no push delivery and refuses rather than inventing an
// event. 501 arrives as kind "unsupported" — a rail declining, not an outage.
try {
  await side.webhook({ body: "{}", headers: {}, now_unix: 1_700_000_000 });
  console.log("webhook     UNEXPECTED: the mock rail produced an event");
} catch (e) {
  const err = e as SidecarHttpError;
  console.log(`webhook     HTTP ${err.status} kind=${err.kind}: ${err.message}`);
}

// The registry is mock-only. Any other rail id is a 404 because this process
// has never heard of it — per-rail registration is unwritten.
try {
  await side.capabilities("solana");
  console.log("registry    UNEXPECTED: a non-mock rail answered");
} catch (e) {
  const err = e as SidecarHttpError;
  console.log(`registry    HTTP ${err.status} kind=${err.kind}: ${err.message}`);
}

// The token gate sits in front of EVERY /v1 route, including the read-only
// capabilities lookup — not just the money-moving ones. A missing header, a
// malformed one and a wrong token are the same detail-free 401.
const bare = await fetch(`${side.baseURL}/v1/rails/mock`);
await bare.body?.cancel();
console.log(`no token    HTTP ${bare.status} on a read-only route — the gate is in front of everything`);
const wrong = await fetch(`${side.baseURL}/v1/rails/mock`, { headers: { Authorization: "Bearer wrong" } });
await wrong.body?.cancel();
console.log(`wrong token HTTP ${wrong.status}, and indistinguishable from the line above`);

File: sdks/kotlin/src/main/kotlin/org/vulos/patala/kotlin/Sidecar.kt
@file:JvmName("PatalaSidecarKt")

package org.vulos.patala.kotlin

import java.time.Duration
import org.vulos.patala.Json
import org.vulos.patala.Patala as JavaPatala
import uniffi.patala.PayRequest

/**
 * A [PayRequest] as the JSON document `patala-sidecar` expects on the wire.
 *
 * The sidecar speaks JSON — that is what an HTTP boundary is — so this path
 * keeps taking and returning strings. What it does not do any more is make you
 * *build* one: [payRequest] produces the same typed record the direct path
 * takes, and this turns it into bytes at the last possible moment.
 *
 * `amountMinor` is written as the `ULong` it is. It is spelled out here rather
 * than handed to a JSON library on purpose: a `Receipt` whose `amount_minor`
 * went through a `Double` is a payments bug that type-checks, and the default
 * number handling of whichever JSON library is already in your build is not
 * this SDK's business.
 */
public fun PayRequest.toJson(): String =
    "{\"amount_minor\":${this.amountMinor}" +
        ",\"currency\":${Json.quote(this.currency)}" +
        ",\"destination\":${Json.quote(this.destination)}" +
        ",\"reference\":${Json.quote(this.reference)}}"

/**
 * Kotlin over [org.vulos.patala.Patala] — patala as a child process on
 * `127.0.0.1`, spoken to over HTTP.
 *
 * ```kotlin
 * PatalaSidecar().use { patala ->
 *     val receipt = patala.charge(payRequest(1250, "USDC", "mock:wallet:alice", "order-4711"))
 *     check(patala.isValid(receipt))
 * }
 * ```
 *
 * `use {}` stops the child process on every path out. Runs on **Java 11+**
 * with no native library, no `--enable-native-access` and no platform matrix
 * — including on Windows, where `libpatala_ffi` has never been built.
 *
 * ### Why you would choose this over the direct path
 *
 * **Key isolation**, and on patala that is the whole argument — not the JVM.
 * A non-custodial rail's signing key lives inside whichever process calls
 * `charge`; route five services through one sidecar and the key lives in one
 * narrow process instead of five. `patala-sidecar/README.md` has the threat
 * model, including what it does not defend against.
 *
 * The *other* products' reason — "loading the shared library replaces the
 * JVM's signal handlers" — is true of llmux and openrate and **false here**.
 * See `README.md`, which carries the measurement rather than the claim.
 *
 * ### There is no `isRefusal(json)` here, and that is the fix
 *
 * This class used to carry `Json.field(verdictJson, "is_refusal") == "true"`,
 * and [org.vulos.patala.Json.field] is a substring scan, not a parser: it did
 * not skip whitespace after the colon. A verdict reformatted anywhere between
 * patala and here read as `" true"`, so `isRefusal` returned **false for a
 * `Malformed` verdict** — failing open, on the one question in this API where
 * failing open loses money.
 *
 * The direct path deleted the identical function rather than repairing it
 * (see [Patala]'s docs — `verdict.isRefusal` is a `Boolean` field decided
 * inside Rust), and [org.vulos.patala.Patala], the Java sidecar client this
 * class wraps under the same no-JSON-dependency constraint, never had one.
 * The sidecar path was simply never migrated. It is now.
 *
 * A better scan is not the fix. patala computes `is_refusal` once, in
 * `DestinationVerdict::is_refusal()`, from an exhaustive match, and puts it on
 * the wire precisely so no binding has to re-derive it. Recovering a boolean
 * from a JSON document is your parser's job — the same parser every other
 * method here hands you a document for — and it must default to refusing:
 *
 * ```kotlin
 * val verdict = mapper.readTree(patala.validateDestination(dest))
 * if (verdict.path("is_refusal").asBoolean(true)) return   // default: refuse
 * ```
 *
 * With no parser on the classpath, `Json.field(verdict, "is_refusal")` will
 * print the field. It is documented for printing, and only that.
 *
 * ### The token is mandatory and this class mints one
 *
 * `patala-sidecar` refuses to start without `PATALA_SIDECAR_TOKEN`. Leaving
 * [token] null mints 32 bytes from `SecureRandom` and sends them as
 * `Authorization: Bearer` on every `/v1` request. `/healthz` is the one
 * unauthenticated route.
 *
 * Pass [token] and [port], or use [PatalaSidecar.attach], to talk to a sidecar
 * somebody else runs — which is the shape key isolation actually takes in
 * production, where one long-lived sidecar serves several services and none of
 * them spawns it.
 */
public class PatalaSidecar private constructor(
    private val delegate: JavaPatala,
    /**
     * The rail id every un-suffixed method below talks to.
     *
     * `patala-sidecar`'s default registry contains exactly one rail, `"mock"`.
     * Any other id is a 404 — not a failure of that rail, but a process that
     * has never heard of it.
     */
    public val railId: String,
) : AutoCloseable {

    public constructor(
        railId: String = JavaPatala.DEFAULT_RAIL,
        port: Int? = null,
        token: String? = null,
        extraEnv: Map<String, String>? = null,
        healthTimeout: Duration = Duration.ofSeconds(20),
        requestTimeout: Duration = Duration.ofSeconds(30),
    ) : this(
        JavaPatala.start(
            JavaPatala.Options().apply {
                this.port = port
                this.token = token
                this.env = extraEnv
                this.timeout = healthTimeout
                this.requestTimeout = requestTimeout
            },
        ),
        railId,
    )

    public companion object {
        /**
         * Talk to a `patala-sidecar` somebody else started. [close] on the
         * result stops nothing.
         */
        public fun attach(
            baseUrl: String,
            token: String,
            railId: String = JavaPatala.DEFAULT_RAIL,
        ): PatalaSidecar = PatalaSidecar(JavaPatala.attach(baseUrl, token), railId)
    }

    /** `http://127.0.0.1:<port>`. */
    public val baseUrl: String get() = delegate.baseUrl()

    /** True when this instance spawned the server it talks to. */
    public val ownsProcess: Boolean get() = delegate.ownsProcess()

    /** `GET /healthz` — the one unauthenticated route. `"ok"`. */
    public fun healthz(): String = delegate.healthz()

    /** `GET /v1/rails/{rail}` — `RailCapabilities`. */
    public fun capabilities(rail: String = railId): String = delegate.capabilities(rail)

    /** `POST /v1/rails/{rail}/quote`. */
    public fun quote(payRequestJson: String, rail: String = railId): String =
        delegate.quote(rail, payRequestJson)

    /** `POST /v1/rails/{rail}/charge` — a `Receipt`. Store it. */
    public fun charge(payRequestJson: String, rail: String = railId): String =
        delegate.charge(rail, payRequestJson)

    /**
     * `POST /v1/rails/{rail}/verify` — `{"valid":true|false}`, **both HTTP
     * 200**. `false` is data, never an HTTP error, precisely so a caller
     * cannot mistake "verified false" for "the sidecar broke".
     */
    public fun verify(receiptJson: String, rail: String = railId): String =
        delegate.verify(rail, receiptJson)

    /** [verify] as a `Boolean`, fail-closed on anything unrecognised. */
    public fun isValid(receiptJson: String, rail: String = railId): Boolean =
        verify(receiptJson, rail).trim() == "{\"valid\":true}"

    /**
     * `POST /v1/rails/{rail}/validate-destination` — the offline pre-flight
     * check. All five verdicts are `200`: a `200` means the rail **answered**,
     * not that the address is good.
     */
    public fun validateDestination(destination: String, rail: String = railId): String =
        delegate.validateDestination(rail, destination)

    /**
     * `POST /v1/rails/{rail}/webhook` — forward the processor's delivery
     * **verbatim**: the exact bytes it sent, and its headers.
     *
     * Every webhook scheme signs the bytes that arrived, so a body that has
     * been through a JSON round-trip on your side is no longer what was
     * signed, and every genuine delivery will fail to authenticate. A rail
     * with no push delivery — the mock — answers 501, which arrives as an
     * exception rather than as an invented event.
     */
    public fun webhook(
        rawBody: ByteArray,
        headers: Map<String, String> = emptyMap(),
        rail: String = railId,
    ): String = delegate.webhook(rail, rawBody, headers)

    /** One raw authenticated GET, for anything this wrapper does not name. */
    public fun get(path: String): String = delegate.get(path)

    /** One raw authenticated POST of a JSON body. */
    public fun post(path: String, body: String): String = delegate.post(path, body)

    /** Stop the child process. Idempotent; `use {}` calls it on every path out. */
    override fun close(): Unit = delegate.close()
}
